modules/consensus/synchronize.go

```diff
@@ -45,121 +42,121 @@ func (g *Gateway) threadedResynchronize() {
 // such transmissions may be required to fully synchronize.
 //
 // TODO: don't run two Synchronize threads at the same time
-func (g *Gateway) Synchronize(peer modules.NetAddress) error {
-	g.log.Println("INFO: synchronizing to", peer)
-	for {
+func (s *State) Synchronize(peer modules.NetAddress) error {
+	// loop until there are no more blocks available
+	moreAvailable := true
+
+	for moreAvailable {
+		// get blockIDs to send
+		id := s.mu.RLock()
+		history := s.blockHistory()
+		s.mu.RUnlock(id)
+
+		// perform RPC
 		var newBlocks []types.Block
-		newBlocks, moreAvailable, err := g.requestBlocks(peer)
+		err := s.gateway.RPC(peer, "SendBlocks", func(conn modules.PeerConn) error {
+			err := encoding.WriteObject(conn, history)
+			if err != nil {
+				return err
+			}
+			err = encoding.ReadObject(conn, &newBlocks, MaxCatchUpBlocks*types.BlockSizeLimit)
+			if err != nil {
+				return err
+			}
+			return encoding.ReadObject(conn, &moreAvailable, 1)
+		})
 		if err != nil {
-			g.log.Printf("ERR: synchronization to %v failed: %v\n", peer, err)
 			return err
 		}
-		g.log.Printf("INFO: %v sent us %v blocks\n", peer, len(newBlocks))
+
+		// integrate received blocks
 		for _, block := range newBlocks {
-			acceptErr := g.state.AcceptBlock(block)
+			// TODO: don't Broadcast these blocks
```

Comment:
we may eventually switch to the bitcoin model, which is to just advertise hashes and leave it up to your peers whether they want to download the full block from you (ditto for txns).

In general though, I think it's a good idea for consensus to know more about the states of its peers.